### PR TITLE
ENH: add apply branch protections tool

### PR DESCRIPTION
**Pull request by tangkong**
## Description
- adds a script to apply branch protections, either for a specific repo or for multiple repos via a csv
- attempts to make this extensible.  Adding a new branch protection rule should only involve modifying the `ProtectionGroup` dataclass and adding a flag in the argparser.
  - some care must be taken to make sure keys in the csv match the dataclass, but something had to match somewhere

## Context
Following up on branch-protecting.

I'd appreciate a careful eye on this, or a test run.  Once I run it, branch protections will be overwritten with no real history or ability to revert.  If we ever did something unique in a branch protection we won't be able to restore it easily.  

There's probably a way to determine the difference between existing rules and the to-be-applied protection ruleset, but in the interest of getting this done I want to punt that.

## Testing
Locally with `python update_branch_protections.py` against pcds-ci-test-repo-python
using subsets of the [branch protections spreadsheet](https://docs.google.com/spreadsheets/d/1y-AAYIK-8awg_SFIGygVQTrAV8KkGF1WJVzbKsLo450/edit?usp=sharing)

Example calls:
`python update_branch_protections.py -h`
`python update_branch_protections.py pcdshub pcds-test-repo-ci-python "Python Dev" --protect-master --protect-pages`
`python update_branch_protections.py --repo-data-path repo_list.csv --write`

<details><summary>example output</summary>
<p>

```bash
(pcds-5.7.3)roberttk@psbuild-rhel7-01:~/src/pcds-python-migration-tools(enh_apply_branch_prot +)$ python update_branch_protections.py --repo-data-path repo_list.csv 
{
  "environments": {
    "nodes": [
      {
        "id": "EN_kwDOIpSrvM4vTCfq",
        "name": "gh-pages",
        "protectionRules": {
          "nodes": []
        }
      },
      {
        "id": "EN_kwDOIpSrvM4vTPsP",
        "name": "github-pages",
        "protectionRules": {
          "nodes": [
            {
              "timeout": 0,
              "reviewers": {
                "nodes": []
              }
            }
          ]
        }
      }
    ]
  },
  "description": "Repository for testing PCDS's continuous integration tools (specifically pcds-ci-helpers)",
  "homepageUrl": "https://pcdshub.github.io/pcds-ci-test-repo-python",
  "name": "pcds-ci-test-repo-python",
  "nameWithOwner": "pcdshub/pcds-ci-test-repo-python",
  "isArchived": false,
  "id": "R_kgDOIpSrvA"
}
(dry run) Deleting branch protection setting
(dry run) Deleting branch protection setting
(dry run) Deleting branch protection setting
created master rule for repo: pcds-ci-test-repo-python
-- rule not applied
created gh_pages rule for repo: pcds-ci-test-repo-python
-- rule not applied
created default rule for repo: pcds-ci-test-repo-python
-- rule not applied
...
```

</p>
</details>

### CODE REVIEW

File: update_branch_protections.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    default: bool = False
+
+    # Map column names to protection rules
+    RULE_MAP = {
+        'master': master_protect,
+        'gh_pages': gh_pages_prot,
+        'default': all_prot
+    }
+
+    # map repo-types to status-check lists
```
Comment (tangkong) [inline update_branch_protections.py:48]: These refer to the [check-lists here](https://github.com/pcdshub/pcds-python-migration-tools/blob/660df8729c601d5da95b8212481053a5897c5e8c/update_github_settings.py#L22-L42)

File: update_branch_protections.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    CHECK_NAME_MAP = {
```
Comment (tangkong) [inline update_branch_protections.py:49]: These refer to the [check-lists here](https://github.com/pcdshub/pcds-python-migration-tools/blob/660df8729c601d5da95b8212481053a5897c5e8c/update_github_settings.py#L22-L42)

File: update_branch_protections.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        'EPICS module': 'none'
+    }
+
+    @classmethod
+    def from_dict(cls, source: dict):
+        try:
+            flags = {f.name: str_to_bool(source[f.name]) for f in fields(cls)
+                     if f.name not in ('owner', 'repo_name', 'repo_type')}
+        except KeyError as e:
+            print("source data malformed, aborting")
```
Comment (klauer) [inline update_branch_protections.py:68]:
```suggestion
            print("source data malformed, aborting data=", source)
```

File: update_branch_protections.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    Create an ArgumentParser for update_branch_protections
+
+    Returns
+    -------
+    argparse.ArgumentParser
+    """
+    parser = argparse.ArgumentParser()
+    # specify an owner, repo, and settings
+    parser.add_argument("owner", type=str, default='pcdshub', nargs='?',
+                        help='Organization or ownder of the repository, "pcdshub"'
```
Comment (klauer) [inline update_branch_protections.py:198]:
```suggestion
                        help='Organization or owner of the repository, "pcdshub"'
```

**Review by klauer — APPROVED** [review]
> I'd appreciate a careful eye on this, or a test run. Once I run it, branch protections will be overwritten with no real history or ability to revert. If we ever did something unique in a branch protection we won't be able to restore it easily.

There are a fair number of PLC projects where I gave up on suggesting pre-commit; those will be affected for sure.
That being said, we just taught people how to use it in the hackathon - so they can no longer claim ignorance.
It'd be good to keep it uniform so I'm still 👍 here

As far as the code goes, looks great to me. I like how simple this layer ends up being with the abstractions underneath.

I expect that you could run this, we could take a look at a few repos, and make sure it looks good. If we spot some errors, we re-evaluate then. Otherwise we wait until any complaints roll in...

File: update_branch_protections.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    parser = argparse.ArgumentParser()
+    # specify an owner, repo, and settings
+    parser.add_argument("owner", type=str, default='pcdshub', nargs='?',
+                        help='Organization or owner of the repository, "pcdshub"'
+                             'by default')
+    parser.add_argument("repo_name", type=str, default='', nargs='?',
+                        help='Name of the repository')
+    parser.add_argument("repo_type", type=str, default='Other', nargs='?',
+                        help='Type of the repository.  Should be one of '
+                             f'{list(ProtectionGroup.CHECK_NAME_MAP.keys())}')
```
Comment (ZLLentz) [inline update_branch_protections.py:204]: I've never tried this before, but I remembered argparse being able to take care of "must be one of these" sorts of args: https://docs.python.org/3/library/argparse.html#choices

File: update_branch_protections.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                        help="Add master protection.  This should be applied as "
+                             "much as possible. This rule requires pull requests and "
+                             "disallows force pushes. This protection also adds required "
+                             "status checks depending on the repo_type")
+    parser.add_argument("--protect-pages", action="store_true", dest='prot_pages',
+                        help="Add gh-pages branch protection, allowing force pushes "
+                             "to the gh-pages branch specifically")
+    parser.add_argument("--protect-default", action="store_true", dest='prot_default',
+                        help="Add default (*) branch protection, disallows "
+                             "the creation of new branches")
```
Comment (ZLLentz) [inline update_branch_protections.py:215]: I know we've agreed to the forking life, but I still expect this to cause some chaos. See the `mfx` repo, for example.

File: update_branch_protections.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+
+    parser.add_argument("--write", action="store_true", dest="write")
+
+    return parser
+
+
+def _main(args=None):
+    """CLI entrypoint."""
+    parser = _create_argparser()
+    return main(**vars(parser.parse_args(args=args)))
```
Comment (ZLLentz) [inline update_branch_protections.py:230]: This is a pretty concise way of consuming the args, I've never seen it before. I like it.

**Review by ZLLentz — APPROVED** [review]
Only minor comments, I think this is good. I went through this carefully as requested, though I haven't given it a test run.

The only thing I'm worried about is fielding questions about why branches can't be created, particularly from people who dislike or don't know about forking.

File: update_branch_protections.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+import csv
+from dataclasses import dataclass, fields
+from pathlib import Path
+from typing import List
+
+from update_github_settings import (BranchProtection, Repository,
+                                    default_required_status_checks)
+
+master_protect = BranchProtection()
+gh_pages_prot = BranchProtection(pattern='gh-pages', allows_force_pushes=True,
```
[the remark this answers is not in the document]
Comment (klauer) [reply update_branch_protections.py:13]: Adding my vote for that formatting style

File: update_branch_protections.py (hunk 6 above)
Comment (klauer) [reply update_branch_protections.py:215]: `mfx` - out of our purview - should probably be set to `exempt`?

Comment (ZLLentz) [reply update_branch_protections.py:215]: possibly, now that it's being managed by the beamline staff themselves

File: update_branch_protections.py (hunk 7 above)
Comment (tangkong) [reply update_branch_protections.py:230]: You can thank Ken for that I shamelessly lifted this from him

File: update_branch_protections.py (hunk 6 above)
Comment (tangkong) [reply update_branch_protections.py:215]: I think as those who have been burdened with maintaining these packages, we also reserve the right to impose our conventions on libraries.  I hope that having well defined workflows with proper guardrails makes contributing easier to understand for newer users.  (I've been meaning to continue writing some more Pull Request disambiguation docs)

I do agree that we can move the specific repos to `exempt` (nnn)  or `other` (ynn, for master protections) as needed.  If anything I hope this tool makes it easier to do so quickly.

File: update_branch_protections.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        return cls(
+            owner=source["owner"],
+            repo_name=source["repo_name"],
+            repo_type=source["repo_type"],
+            **flags
+        )
+
+    def apply_protections(self, write: bool) -> None:
+        """
+        Create a ``Repository``, delete its existing branch protections, and
```
[the remark this answers is not in the document]
Comment (tangkong) [reply update_branch_protections.py:80]: Ah that's fair, I can clarify this

File: update_branch_protections.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tangkong) [reply update_branch_protections.py:13]: I usually avoid the dedented style when I can fit more than one arg per line, making the line difference more extreme.  But there's no reason not to use it here 👍

File: update_branch_protections.py (hunk 7 above)
Comment (ZLLentz) [reply update_branch_protections.py:230]: I guess I probably have seen it before then, and the information leaked out of my skull!

**Pull request merged**